### ISSUE TITLE
DPL-848 As a TL (Tom) I would like a RabbitMQ producer to publish events about the sequencing run status for samples so that applications can subscribe to updates without polling the MLWH database.

### ISSUE THREAD

**Issue by TWJW-SANGER** [sanger/General-Backlog-Items#331]
**User story**
As a TL (Tom) I would like a RabbitMQ producer to publish events about the sequencing run status for samples so that applications can subscribe to updates without polling the MLWH database.

**Who are the primary contacts for this story**
TW

**Who is the nominated tester for UAT**
PSD team review and test only

**Acceptance criteria**
To be considered successful design should:
- [x] Create a RabbitMQ fan out exchange on the Platform Solutions RabbitMQ system for UAT,Training and Production.
- [x] Consider how the deployment project interacts with RabbitMQ for setup
- [x] Create a service that Polls MLWH database every 5 to 30 mins for sequencing run_status changes made since the last run.
- [x] Create a Message Schema to define the broadcast message being sent.
- [ ] Register the Message Schema with RedPanda service
- [x] Document/Record message schema's and how to use them on confluence – anticipate there will be more queries
- [ ] The service can be deployed to our training,  uat & production environments on openstack.

**Additional context**
See Confluence page - [How to query for all sequencing run status changes for a given study after a given time point](https://ssg-confluence.internal.sanger.ac.uk/display/PSDPUB/How+to+query+for+all+sequencing+run+status+changes+for+a+given+study+after+a+given+time+point)

Consider what already exists in front of MLWH - there are RabbitMQ consumers of some kind - do we want to reuse that platform? 

**Blocked By**
#332

**Comment by emrojo** [comment]
I think this process should be polling MLWH for new runs, and then publishing events in the event warehouse queues, so:
- The exchange is already defined there 'psd.sequencescape'
- The queue is not needed, all will be handled by the event warehouse consumer currently running there
- The routing key must follow the format: staging.event.# for uat and production.event.# for production
- The message schema is kind of defined in the readme in here: https://github.com/sanger/event_warehouse

**Comment by TWJW-SANGER** [comment]
I'm not sure but there is some advice copied from [here](https://groups.google.com/g/rabbitmq-discuss/c/Jp49IRe693o) below:

_1+ year and several apps later I see the following pattern developing on my team:_

1. Exchanges are declared by producers.
2. Queues are declared and bound by consumers.
3. In vast majority of cases, consumers are "smart" and producers are very dumb: they just put messages in well-known places. Consumers
    are "smart" in a sense that they almost always keep some persistent state (in Redis, for example), while producers are just monitored in case they crash.
4. Exchanges are named using the following pattern $prefix.$applications_group.$app.$environment. For example, [megacorp.mail_delivery.shaper.qa](http://megacorp.mail_delivery.shaper.qa/) or megacorp.search.online_indexer.production. It works well with multiple installations and if you want to route some data to your customers/partners/archive


For us this exchange is about broadcasting update for the sequencing status to a sample to whoever is interested.
So in the spirit of the above, I would suggest:

Exchange name: sanger.psd.sample_status.$environment where $environment is one of prod, uat, etc...




Does something similar work for queues? Although I believe these are up to the consumers?

Routing key - Does a fanout exchange look at this?

For the message schema - sample_sequence_status ?


I'd welcome any comments or critique, given zero experience with RabbitMQ myself.

**Comment by javidahmed64592** [comment]
I have began creating a PR for my progress so far. I have added a new task to [janitor](https://github.com/sanger/janitor) called `sequencing_publisher` which polls MLWH for sample sequence run status changes. The query currently returns an empty list so this needs to be investigated further.

Once that is fixed, the results from the query will be parsed into messages ready to be sent to `RabbitMQ` to an exchange named `sanger.psd.sample_status.uat` using the `sample_sequence_status.avsc` message schema file. This is already in place but the schema needs to be registered to Redpanda and publishing needs to be tested once the aforementioned query is able to return the desired results.

I still need to write unit tests and update relevant documentation, then update the `deployment` project. It should be noted that currently, [janitor](https://github.com/sanger/janitor) has not been deployed to the `training` environment since there is no MLWH database there so this should be taken into consideration and potentially discussed in further detail.
